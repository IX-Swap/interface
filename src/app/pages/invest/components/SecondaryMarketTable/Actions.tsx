import React from 'react'
import { Launch as LaunchIcon } from '@material-ui/icons'
import { Button, IconButton } from '@material-ui/core'
import { AppRouterLinkComponent } from 'components/AppRouterLink'
import { InvestRoute as paths } from 'app/pages/invest/router/config'
import { OTCMarketRoute } from 'app/pages/exchange/router/config'
import { useTheme } from '@material-ui/core/styles'

export interface ActionsProps {
  item: any
}

export const Actions = ({ item }: ActionsProps) => {
  const theme = useTheme()
  return (
    <>
      <IconButton
        component={AppRouterLinkComponent}
        to={paths.viewListing}
        params={{
          userId: item.listing.createdBy,
          listingId: item.listing._id
        }}
        size='small'
      >
        <LaunchIcon color='disabled' />
      </IconButton>
      <Button
        style={{
          marginLeft: theme.spacing(4.5),
          marginRight: theme.spacing(4),
          fontSize: 14,
          color: theme.palette.slider.activeBackground
        }}
        component={AppRouterLinkComponent}
        to={OTCMarketRoute.market}
        params={{
          pairId: item._id
        }}
        size='small'
      >
        TRADE
      </Button>
    </>
  )
}
