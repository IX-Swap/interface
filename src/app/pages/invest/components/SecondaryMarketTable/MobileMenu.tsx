import { Box, Drawer, Grid, IconButton, Typography } from '@mui/material'
import { useStyles } from 'app/pages/accounts/pages/banks/pages/BanksList/MobileMenu.styles'
import React, { useContext, useMemo } from 'react'
import { Icon } from 'ui/Icons/Icon'
import { ActiveElementContext } from 'app/context/ActiveElementContextWrapper'
import { NewAction } from 'app/pages/authorizer/components/NewAction'
import { InvestRoute } from 'app/pages/invest/router/config'
import { Launch as LaunchIcon } from '@mui/icons-material'
import { history } from 'config/history'
import { generatePath } from 'react-router-dom'
import { ReactComponent as TradeIcon } from './trade-icon.svg'

export interface MobileMenuProps {
  items: any[]
}

export const MobileMenu = (props: MobileMenuProps) => {
  const context = useContext(ActiveElementContext)
  const open = context?.hasOpenIndices
  const classes = useStyles()
  const openIndex = context?.openIndex

  const { items } = props

  const selectedItem = useMemo(
    () => items.filter(item => item._id === openIndex)?.[0],
    [openIndex, items]
  )
  if (selectedItem === undefined) {
    return null
  }

  const actionsPropsList = [
    {
      label: 'Learn More',
      icon: <LaunchIcon color='disabled' />,
      onClick: () =>
        history.push(
          generatePath(InvestRoute.viewListing, {
            userId: selectedItem.listing.createdBy,
            listingId: selectedItem.listing._id
          })
        )
    },
    {
      label: 'Trade',
      icon: <TradeIcon />,
      onClick: () =>
        history.push(
          generatePath(InvestRoute.exchange, {
            pairId: selectedItem._id
          })
        )
    }
  ]

  const onClose = () => {
    if (openIndex !== undefined) {
      context?.toggleRow(openIndex)
    }
  }

  return (
    <>
      <Drawer
        anchor='bottom'
        open={open}
        onClose={onClose}
        PaperProps={{ sx: { backgroundColor: 'transparent' } }}
      >
        <Box className={classes.drawer}>
          <Box className={classes.titleRow}>
            <Typography
              variant='subtitle2'
              fontWeight={600}
              className={classes.header}
            >
              {selectedItem.name}
            </Typography>
            <IconButton onClick={onClose} size='medium'>
              <Icon name='close' />
            </IconButton>
          </Box>
          <Grid flexDirection='column'>
            <Grid display='flex' flexDirection={'column'} gap={2}>
              {actionsPropsList.map(actionProps => (
                <NewAction {...actionProps} />
              ))}
            </Grid>
            <Box mb={2} />
          </Grid>
        </Box>
      </Drawer>
    </>
  )
}
