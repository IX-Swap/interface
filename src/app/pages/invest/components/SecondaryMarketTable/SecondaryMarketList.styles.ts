import makeStyles from '@mui/styles/makeStyles'

export const useStyles = makeStyles(theme => ({
  cell: {
    padding: theme.spacing(2, 0, 0)
  },
  card: {
    backgroundColor: theme.palette.background.paper,
    padding: theme.spacing(3),
    borderRadius: 8
  },
  columns: {
    display: 'grid',
    gridTemplateColumns: '1fr 1fr',
    rowGap: theme.spacing(3),
    columnGap: theme.spacing(2)
  },
  additionalColumns: {
    display: 'grid',
    gridTemplateColumns: '1fr',
    rowGap: theme.spacing(3),
    columnGap: theme.spacing(2)
  },
  label: {
    color: theme.palette.text.secondary
  },
  value: {
    fontSize: 14
  },
  iconButton: {
    background: theme.palette.button.bgLight
  }
}))
