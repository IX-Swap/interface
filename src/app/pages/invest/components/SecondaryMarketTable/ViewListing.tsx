import { ListingDetails } from 'app/pages/exchange/components/ListingDetails/ListingDetails'
import { useListing } from 'app/pages/exchange/hooks/useListing'
import React from 'react'

export const ViewListing = () => {
  const { data, isLoading } = useListing()

  if (isLoading || data === undefined) {
    return null
  }

  return <ListingDetails withoutActions={true} data={data} />
}
