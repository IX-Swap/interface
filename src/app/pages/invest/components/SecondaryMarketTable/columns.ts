import { formatMoney } from 'helpers/numbers'
import { TableColumn } from 'types/util'
import { getUserNameById } from 'helpers/tables'

export interface Order {
  createdAt: string
  pair: string
  side: string
  price: number
  amount: number
  total: number
  filled: number
}

export const columns: Array<TableColumn<any>> = [
  {
    key: '',
    label: ''
  },
  {
    key: 'name',
    label: 'Pair'
  },
  {
    label: 'Symbol',
    key: 'listing.tokenSymbol'
  },
  {
    key: 'listing.tokenName',
    label: 'Name'
  },
  {
    key: 'listing.createdBy',
    label: 'Issued By',
    align: 'center',
    headAlign: 'center',
    render: (_, value) => getUserNameById(value.listing.createdBy)
  },

  {
    key: 'listing.minimumTradeUnits',
    label: 'Price',
    align: 'right',
    headAlign: 'right',
    render: (_, value) => formatMoney(value.listing.minimumTradeUnits, '')
  },
  {
    key: '',
    label: ''
  },
  {
    key: 'listing.marketType',
    label: 'Type'
  }
]
