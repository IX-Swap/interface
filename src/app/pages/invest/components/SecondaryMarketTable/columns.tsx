import React from 'react'
import { Box } from '@mui/material'
import { TableColumn } from 'types/util'
import { formatMoney } from 'helpers/numbers'
import { useAppBreakpoints } from 'hooks/useAppBreakpoints'
import { useQueryFilter } from 'hooks/filters/useQueryFilter'
import { ReactComponent as SortActiveIcon } from './icons/sort_active.svg'
import { ReactComponent as SortDefaultIcon } from './icons/sort_default.svg'

export interface Order {
  createdAt: string
  pair: string
  side: string
  price: number
  amount: number
  total: number
  filled: number
}

export const PriceLabelWithSort = () => {
  const { isTablet } = useAppBreakpoints()
  const { getFilterValue, updateFilter, removeFilter } = useQueryFilter()
  const sortValue = getFilterValue('sortBy')
  const onClick = () => {
    switch (sortValue) {
      case undefined:
        updateFilter('sortBy', 'DESC')
        break
      case 'DESC':
        updateFilter('sortBy', 'ASC')
        break
      default:
        removeFilter('sortBy')
        break
    }
  }

  if (isTablet) {
    return <>Price</>
  }

  return (
    <Box
      onClick={onClick}
      display={'flex'}
      alignItems={'center'}
      style={{ cursor: 'pointer', userSelect: 'none' }}
    >
      Price
      <Box
        display={'flex'}
        flexDirection={'column'}
        style={{
          transform: `rotate(${sortValue === 'ASC' ? 0 : '180deg'})`
        }}
      >
        {sortValue === undefined ? <SortDefaultIcon /> : <SortActiveIcon />}
      </Box>
    </Box>
  )
}

export const renderLatestPrice = (val: number, row: any) => {
  return (
    <Box width={'100%'} height={'100%'}>
      {formatMoney(val, row.name.split('/')[1], true)}
    </Box>
  )
}

export const renderCreatedBy = (_: string, row: any) => row.listing.user.name

export const columns: Array<TableColumn<any>> = [
  {
    key: 'name',
    label: <Box sx={{ paddingLeft: { xs: 0, lg: 2 } }}>Pair</Box>
  },
  {
    label: 'Symbol',
    key: 'listing.tokenSymbol'
  },
  {
    key: 'listing.tokenName',
    label: 'Name'
  },
  {
    key: 'listing.createdBy',
    label: 'Issued By',
    align: 'left',
    headAlign: 'left',
    render: renderCreatedBy
  },

  {
    key: 'latestPrice',
    label: <PriceLabelWithSort />,
    align: 'left',
    headAlign: 'left',
    render: renderLatestPrice
  },
  {
    key: 'listing.marketType',
    label: 'Type'
  }
]
