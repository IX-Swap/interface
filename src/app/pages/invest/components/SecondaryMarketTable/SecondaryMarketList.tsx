import { Button, Grid, TableBody, TableCell, TableRow } from '@mui/material'
import { ActiveElementContext } from 'app/context/ActiveElementContextWrapper'
import { useStyles } from './SecondaryMarketList.styles'
import { CompactRowProps } from 'components/TableWithPagination/CompactRow'
import get from 'lodash/get'
import React, { useContext } from 'react'
import { TableColumn } from 'types/util'
import { Icon } from 'ui/Icons/Icon'
import { TableViewRendererProps } from 'ui/UIKit/TablesKit/components/TableView/TableView'
import { MobileMenu } from 'app/pages/invest/components/SecondaryMarketTable/MobileMenu'
import { NoOffers } from 'app/pages/invest/components/NoOffers/NoOffers'
import { useQueryCache } from 'react-query'

export interface CompactBodyProps<T> extends TableViewRendererProps<T> {
  renderRow?: (props: CompactRowProps<T>) => JSX.Element
}

export interface RenderCellProps extends TableColumn<any, string> {
  item: any
}

export const renderCell = ({ render, key, item }: RenderCellProps) => {
  return (
    key.length > 0 &&
    (typeof render === 'function'
      ? render(get(item, key), item)
      : get(item, key))
  )
}

export const SecondaryMarketList = (props: CompactBodyProps<any>) => {
  const queryCache = useQueryCache()
  const { columns, items, cacheQueryKey } = props

  const classes = useStyles()
  const context = useContext(ActiveElementContext)

  const onButtonClick = (item: any) => {
    context?.toggleRow(item._id)
  }

  if (
    queryCache.watchQuery(cacheQueryKey).getCurrentResult().status === 'loading'
  ) {
    return null
  }

  if (items.length < 1) {
    return <NoOffers />
  }

  const renderColumnsBlock = (
    columns: any[],
    item: any,
    fullWidth: boolean = false
  ) => {
    return (
      <Grid
        item
        container
        rowGap={1}
        className={fullWidth ? classes.additionalColumns : classes.columns}
        xs={12}
      >
        {columns.map(({ label, key, render }) => (
          <Grid item container key={key} flexDirection='row' rowGap={1}>
            <Grid item xs={12} className={classes.label}>
              {label}
            </Grid>
            <Grid item xs={12} className={classes.value}>
              {renderCell({ render, key, item, label })}
            </Grid>
          </Grid>
        ))}
      </Grid>
    )
  }

  return (
    <>
      <TableBody>
        {items.map((item, i) => (
          <TableRow key={i}>
            <TableCell className={classes.cell}>
              <Grid
                container
                flexDirection='row'
                rowGap={3}
                className={classes.card}
              >
                {renderColumnsBlock(columns.slice(0, 2), item)}
                {renderColumnsBlock(columns.slice(2, 4), item, true)}
                {renderColumnsBlock(columns.slice(4, columns.length), item)}
                <Grid item xs={12}>
                  <Button
                    onClick={() => onButtonClick(item)}
                    fullWidth
                    className={classes.iconButton}
                  >
                    <Icon name='more-horizontal' />
                  </Button>
                </Grid>
              </Grid>
            </TableCell>
          </TableRow>
        ))}
      </TableBody>
      <MobileMenu items={items} />
    </>
  )
}
