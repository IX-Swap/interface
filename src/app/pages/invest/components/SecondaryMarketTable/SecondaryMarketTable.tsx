import React from 'react'
import { Button, Grid } from '@material-ui/core'
import { TableView } from 'components/TableWithPagination/TableView'
import { exchange as exchangeQueryKeys } from 'config/queryKeys'
import { useQueryFilter } from 'hooks/filters/useQueryFilter'
import { Actions } from 'app/pages/invest/components/SecondaryMarketTable/Actions'
import { exchange as exchangeURL } from 'config/apiURL'
import { columns } from 'app/pages/invest/components/SecondaryMarketTable/columns'
import { SearchFilter } from 'app/components/SearchFilter'
import { AppRouterLinkComponent } from 'components/AppRouterLink'
import { useTheme } from '@material-ui/core/styles'
import { OTCMarketRoute } from 'app/pages/exchange/router/config'

export const SecondaryMarketTable = () => {
  const theme = useTheme()
  const { getFilterValue } = useQueryFilter()
  const search = getFilterValue('search')
  const secondaryMarketSearch = getFilterValue('secondaryMarketSearch')

  return (
    <Grid container direction='column' spacing={3}>
      <Grid item container justify={'space-between'}>
        <Grid item xs={3}>
          <SearchFilter
            data-testid='secondaryMarketSearch'
            fullWidth
            filterValue={'secondaryMarketSearch'}
            inputAdornmentPosition='end'
            placeholder='Search'
          />
        </Grid>
        <Grid item>
          <Button
            component={AppRouterLinkComponent}
            to={OTCMarketRoute.holdings}
            color='primary'
            variant='outlined'
            size='large'
            disableElevation
            style={{ color: theme.palette.primary.main }}
          >
            My Holdings
          </Button>
        </Grid>
      </Grid>
      <Grid item>
        <TableView
          uri={exchangeURL.marketList}
          name={exchangeQueryKeys.marketList}
          columns={columns}
          hasActions
          actions={Actions}
          filter={{ listingKeyword: search ?? secondaryMarketSearch } as any}
        />
      </Grid>
    </Grid>
  )
}
