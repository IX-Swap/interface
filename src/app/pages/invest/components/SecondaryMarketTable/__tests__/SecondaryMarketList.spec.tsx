import { ActiveElementContextWrapper } from 'app/context/ActiveElementContextWrapper'
import { exchange as exchangeQueryKeys } from 'config/queryKeys'
import React from 'react'
import { render } from 'test-utils'
import { SecondaryMarketList } from 'app/pages/invest/components/SecondaryMarketTable/SecondaryMarketList'
import { listing } from '__fixtures__/listings'
import { columns } from '../columns'

jest.mock('app/pages/accounts/pages/banks/pages/BanksList/MobileMenu', () => ({
  MobileMenu: jest.fn(() => null)
}))

describe('SecondaryMarketList', () => {
  const props = {
    columns: columns,
    items: [{ name: 'test', latestPrice: 100, listing: listing }],
    loading: false,
    hasActions: false,
    cacheQueryKey: exchangeQueryKeys.marketList
  }

  it('compact bank list matches snapshot', () => {
    const { container } = render(
      <ActiveElementContextWrapper>
        <SecondaryMarketList {...props} />
      </ActiveElementContextWrapper>
    )
    expect(container).toMatchSnapshot()
  })
})
