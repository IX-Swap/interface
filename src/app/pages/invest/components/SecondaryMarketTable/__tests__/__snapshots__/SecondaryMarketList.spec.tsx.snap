// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecondaryMarketList compact bank list matches snapshot 1`] = `<div />`;
